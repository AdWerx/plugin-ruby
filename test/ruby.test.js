#!/usr/bin/env node

const { spawn, spawnSync } = require("child_process");
const fs = require("fs");
const path = require("path");
const prettier = require("prettier");

const print = require("../src/print");
const nodes = require("../src/nodes");
const escapePattern = require("../src/escapePattern");

const expectedMinitestFiles = [
  "alias.rb",
  "binary.rb",
  "kwargs.rb",
  "regexp.rb"
];

const expectedUnhandledNodes = [
  "arg_ambiguous",
  "args_add",
  "args_new",
  "heredoc_dedent",
  "magic_comment",
  "mlhs_add",
  "mlhs_new",
  "mrhs_add",
  "mrhs_new",
  "operator_ambiguous",
  "parse_error",
  "qsymbols_add",
  "qsymbols_new",
  "qwords_add",
  "qwords_new",
  "regexp_add",
  "regexp_new",
  "stmts_add",
  "stmts_new",
  "string_add",
  "string_content",
  "symbols_add",
  "symbols_new",
  "void_stmt",
  "words_add",
  "words_new",
  "xstring_add",
  "xstring_new",
  ""
];

const format = (file, config) => prettier.format(fs.readFileSync(file, "utf8"), {
  parser: "ruby", plugins: ["."], ...config
});

const eachConfig = callback => fs.readdirSync("./test/config").forEach(prettierConfig => {
  if (prettierConfig.match(/.+\.json$/)) {
    const rubocopConfig = `./test/config/${prettierConfig.slice(0, -5)}.yml`;
    const config = JSON.parse(fs.readFileSync(`./test/config/${prettierConfig}`, "utf8"));
    callback(prettierConfig, rubocopConfig, config);
  }
});

const eachTest = (config, callback) => fs.readdirSync("./test/cases").forEach(file => {
  callback(file, () => format(`./test/cases/${file}`, config));
});

const eachError = (config, callback) => fs.readdirSync("./test/errors").forEach(file => {
  callback(file, () => format(`./test/errors/${file}`, config));
});

const eachUnsupportedNode = callback => {
  const child = spawnSync("ruby", [
    "-rripper", "-e", "puts Ripper::PARSER_EVENTS"
  ]);

  const error = child.stderr.toString();
  if (error) {
    throw new Error(error);
  }

  const events = Object.keys(nodes).concat(expectedUnhandledNodes);
  child.stdout.toString().split("\n").forEach(event => {
    if (events.indexOf(event) === -1) {
      callback(event);
    }
  });
};

const handleChildProcess = child => new Promise((resolve, reject) => (
  child.on("exit", code => {
    if (code === 0) {
      resolve();
    } else {
      reject((child.stdout.read() || child.stderr.read() || "").toString());
    }
  })
));

let tmpDir;

beforeAll(() => {
  tmpDir = fs.mkdtempSync("minitest-");
});

afterAll(() => {
  fs.readdirSync(tmpDir).forEach(file => {
    fs.unlinkSync(path.join(tmpDir, file));
  });

  fs.rmdirSync(tmpDir);
});

eachConfig((prettierConfig, rubocopConfig, config) => {
  describe(`for the ${prettierConfig} config`, () => {
    eachTest(config, (file, getContents) => {
      describe(file, () => {
        test(`matches expected output`, () => {
          expect(getContents()).toMatchSnapshot();
        });

        if (!process.env.NOLINT) {
          test(`generated code passes rubocop`, () => {
            const child = spawn("bundle", [
              "exec", "rubocop", "--stdin", file, "--config", rubocopConfig
            ]);

            if (process.env.VIOLATIONS) {
              child.stdout.pipe(process.stdout);
            }

            child.stdin.write(getContents());
            child.stdin.end();

            return handleChildProcess(child);
          });
        }

        if (expectedMinitestFiles.includes(file)) {
          test(`generated code passes as a ruby test`, () => {
            const filepath = path.join(tmpDir, file);
            fs.writeFileSync(filepath, getContents());

            const child = spawn("bundle", [
              "exec", "ruby", "test/minitest.rb", tmpDir, file
            ]);

            return handleChildProcess(child);
          });
        } else {
          test.todo(`generated code passes as a ruby test`);
        }
      });
    });

    eachError(config, (file, getContents) => {
      test(`${file} throws error on parsing`, () => {
        expect(getContents).toThrowError();
      });
    });
  });
});

describe("node support", () => {
  eachUnsupportedNode(event => {
    test(`handles the ${event} event`, () => {
      expect(event).toEqual("handled");
    });
  });
});

test("when encountering an unsupported node type", () => {
  const path = { getValue: () => ({ type: "unsupported", body: {} }) };

  expect(() => print(path)).toThrow("Unsupported");
});

describe("escape sequences", () => {
  const should = value => string => (
    expect(escapePattern.test(string)).toBe(value)
  );

  const shouldMatch = should(true);
  const shouldNotMatch = should(false);

  test("identifies simple escapes", () => {
    shouldMatch("\\t");
    shouldMatch("\\n");
    shouldNotMatch("\\x");
  });

  test("identifies octal bits", () => {
    shouldMatch("\\123");
    shouldNotMatch("\\1a3");
  });

  test("identifies hex bits", () => {
    shouldMatch("\\xab");
    shouldNotMatch("\\xag");
  });

  test("identifies unicode char", () => {
    shouldMatch("\\uabcd");
    shouldNotMatch("\\uabcg");
  });

  test("identifies unicode chars", () => {
    shouldMatch("\\u{abcd abce abcf}");
    shouldNotMatch("\\u{abcd abce abcg}");
  });

  test("identifies controls", () => {
    shouldMatch("\\ca");
    shouldMatch("\\C-a");
    shouldNotMatch("\\cðŸŽ‰");
  });

  test("identifies metas", () => {
    shouldMatch("\\M-a");
    shouldNotMatch("\\M-ðŸŽ‰");
  });

  test("identifies meta controls", () => {
    shouldMatch("\\M-\\C-a");
    shouldMatch("\\M-\\ca");
    shouldMatch("\\c\\M-a");
  });

  test("identifies deletes", () => {
    shouldMatch("\\c?");
    shouldMatch("\\C-?");
    shouldNotMatch("\\d?");
  });
});
