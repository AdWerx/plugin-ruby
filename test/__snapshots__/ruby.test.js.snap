// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alias.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

alias a b # first

alias c d

alias $e $f # rubocop:disable Style/GlobalVars
"
`;

exports[`alias.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

alias a b # first

alias c d

alias $e $f # rubocop:disable Style/GlobalVars
"
`;

exports[`array.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

[]

[1, 2, 3]

[\\"a\\", \\"b\\", \\"c\\"]

%w[a b c]

%i[a b c]

%W[a#{a}a b#{b}a c#{c}c]

%I[a#{a}a b#{b}b c#{c}c]

[1, 2, *[3, 4], 5, 6]

[
  super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_long, [
    super_super_super_super_super_super_super_super_super_super_super_long,
    super_super_super_super_super_super_super_super_super_super_super_long
  ]
]

a[1]

a[
  super_super_super_super_super_super_super_super_super_super_super_super_sulong
]

a[1] = 2

a[
  super_super_super_super_super_super_super_super_super_super_super_super_sulong
] =
  super_super_super_super_super_super_super_super_super_super_super_super_long

a[1] = [
  super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_long
]

# rubocop:enable Lint/Void
"
`;

exports[`array.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

[]

[1, 2, 3]

['a', 'b', 'c']

%w[a b c]

%i[a b c]

%W[a#{a}a b#{b}a c#{c}c]

%I[a#{a}a b#{b}b c#{c}c]

[1, 2, *[3, 4], 5, 6]

[
  super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_long, [
    super_super_super_super_super_super_super_super_super_super_super_long,
    super_super_super_super_super_super_super_super_super_super_super_long
  ]
]

a[1]

a[
  super_super_super_super_super_super_super_super_super_super_super_super_sulong
]

a[1] = 2

a[
  super_super_super_super_super_super_super_super_super_super_super_super_sulong
] =
  super_super_super_super_super_super_super_super_super_super_super_super_long

a[1] = [
  super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_long
]

# rubocop:enable Lint/Void
"
`;

exports[`assign.rb matches expected output for off.json 1`] = `
"a = 1

a =
  begin
    2
  end

a, b, c = [1, 2, 3]

a = 1, 2, 3

a, b, c = 1, 2, 3

a, *b = 1, 2, 3

a, *b, c, d = 1, 2, 3

a, * = 1, 2, 3

a = *[1, 2, 3]

(a, b), c = [1, 2], 3

* = [1, 2, 3]

*, a = [1, 2, 3]

super_super_super_super_super_long, super_super_super_super_super_long, super_super_super_super_super_long =
  super_super_super_super_super_long,
  super_super_super_super_super_long,
  super_super_super_super_super_long

a ||= 1

a ||=
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long

a = [
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
]

a = {
  :a =>
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  :b =>
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  :c =>
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
}

a = [
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
].sort

a = {
  :a =>
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  :b =>
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
}.sort

a.a = 1

super_super_super_super_super_super_super_long.super_super_super_super_super_super_super_long =
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
"
`;

exports[`assign.rb matches expected output for on.json 1`] = `
"a = 1

a =
  begin
    2
  end

a, b, c = [1, 2, 3]

a = 1, 2, 3

a, b, c = 1, 2, 3

a, *b = 1, 2, 3

a, *b, c, d = 1, 2, 3

a, * = 1, 2, 3

a = *[1, 2, 3]

(a, b), c = [1, 2], 3

* = [1, 2, 3]

*, a = [1, 2, 3]

super_super_super_super_super_long, super_super_super_super_super_long, super_super_super_super_super_long =
  super_super_super_super_super_long,
  super_super_super_super_super_long,
  super_super_super_super_super_long

a ||= 1

a ||=
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long

a = [
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
]

a = {
  a:
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  b:
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  c:
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
}

a = [
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
].sort

a = {
  a:
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long,
  b:
    super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
}.sort

a.a = 1

super_super_super_super_super_super_super_long.super_super_super_super_super_super_super_long =
  super_super_super_super_super_super_super_super_super_super_super_super_super_super_long
"
`;

exports[`binary.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

a && b && c && d

super_super_super_super_super_long && super_super_super_super_super_long &&
  super_super_super_super_super_long
"
`;

exports[`binary.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

a && b && c && d

super_super_super_super_super_long && super_super_super_super_super_long &&
  super_super_super_super_super_long
"
`;

exports[`blocks.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/UnusedBlockArgument
# We should fix this to automatically prefix the variables with _.

loop { 1 }

loop { 1 }

loop do
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

loop { super_super_super_super_super_super_super_super_super_super_super_long }

loop { |i| 1 }

loop { |i| i }

loop { |*| i }

loop { |(a, b)| i }

loop { |a, (b, c), d, *e| i }

loop do |i|
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

loop { |i| super_super_super_super_super_super_super_super_super_super_long }

loop(&:i)

loop(&:i)

loop { |i, j| i.to_s }

[1, 2, 3].each do |i|
  p i
end

# rubocop:disable Lint/AmbiguousBlockAssociation
# We should handle this.

target.method object.map { |arg| arg * 2 }

# rubocop:enable Lint/UnusedBlockArgument, Lint/AmbiguousBlockAssociation
"
`;

exports[`blocks.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/UnusedBlockArgument
# We should fix this to automatically prefix the variables with _.

loop { 1 }

loop { 1 }

loop do
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

loop { super_super_super_super_super_super_super_super_super_super_super_long }

loop { |i| 1 }

loop { |i| i }

loop { |*| i }

loop { |(a, b)| i }

loop { |a, (b, c), d, *e| i }

loop do |i|
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

loop { |i| super_super_super_super_super_super_super_super_super_super_long }

loop(&:i)

loop(&:i)

loop { |i, j| i.to_s }

[1, 2, 3].each do |i|
  p i
end

# rubocop:disable Lint/AmbiguousBlockAssociation
# We should handle this.

target.method object.map { |arg| arg * 2 }

# rubocop:enable Lint/UnusedBlockArgument, Lint/AmbiguousBlockAssociation
"
`;

exports[`break.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

[].each { break }

[].each { break 1 }

[].each { break(1) }

[].each { break 1, 2 }
"
`;

exports[`break.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

[].each { break }

[].each { break 1 }

[].each { break(1) }

[].each { break 1, 2 }
"
`;

exports[`case.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/EmptyWhen
# Disabling EmptyWhen on account of
# https://github.com/rubocop-hq/rubocop/issues/3696.

# rubocop:disable Style/EmptyCaseCondition
# This is something that we should handle by converting each of the when
# statements into valid if/elsif/else statements.

case
when a
  1
end

case a
when b
  1
end

case a
when b, c
  1
end

case a
when b
when c
  1
end

case a
when b
  1
when c
  2
end

case a
when b
  1
else
  2
end

# rubocop:enable Lint/EmptyWhen, Style/EmptyCaseCondition
"
`;

exports[`case.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/EmptyWhen
# Disabling EmptyWhen on account of
# https://github.com/rubocop-hq/rubocop/issues/3696.

# rubocop:disable Style/EmptyCaseCondition
# This is something that we should handle by converting each of the when
# statements into valid if/elsif/else statements.

case
when a
  1
end

case a
when b
  1
end

case a
when b, c
  1
end

case a
when b
when c
  1
end

case a
when b
  1
when c
  2
end

case a
when b
  1
else
  2
end

# rubocop:enable Lint/EmptyWhen, Style/EmptyCaseCondition
"
`;

exports[`class.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

module Pret
  module Tier
    class Object; end

    class Object; end

    class Object
      attr_accessor :foo
    end

    class Object < BasicObject; end

    class Object < BasicObject; end

    class Object < BasicObject
      attr_accessor :bar
    end

    class << self
      def method; end

      undef method
    end
  end
end

Pret::Tier::Object # rubocop:disable Lint/Void
Pret::TIER = \\"config\\".freeze

::Pret::Tier::Object # rubocop:disable Lint/Void
::PRET = \\"config\\".freeze
"
`;

exports[`class.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

module Pret
  module Tier
    class Object; end

    class Object; end

    class Object
      attr_accessor :foo
    end

    class Object < BasicObject; end

    class Object < BasicObject; end

    class Object < BasicObject
      attr_accessor :bar
    end

    class << self
      def method; end

      undef method
    end
  end
end

Pret::Tier::Object # rubocop:disable Lint/Void
Pret::TIER = 'config'.freeze

::Pret::Tier::Object # rubocop:disable Lint/Void
::PRET = 'config'.freeze
"
`;

exports[`defined.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

defined?(a) # first

defined?(a) # second

defined?(
  super_super_super_super_super_super_super_super_super_super_super_super_long
)

defined?(
  super_super_super_super_super_super_super_super_super_super_super_super_long
)

defined?(
  super_super_super_super_super_super_super_super_super_super_super_super_long
)

defined?(a)

# rubocop:enable Lint/Void # third
"
`;

exports[`defined.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

defined?(a) # first

defined?(a) # second

defined?(
  super_super_super_super_super_super_super_super_super_super_super_super_long
)

defined?(
  super_super_super_super_super_super_super_super_super_super_super_super_long
)

defined?(
  super_super_super_super_super_super_super_super_super_super_super_super_long
)

defined?(a)

# rubocop:enable Lint/Void # third
"
`;

exports[`hash.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

{}

{ :a => \\"a\\", :b => \\"b\\", :c => \\"c\\" }

{ :a => \\"a\\", :b => \\"b\\", :c => \\"c\\" }

{ Foo => 1, Bar => 2 }

{
  :super_super_super_super_super_super_super_super_long =>
    super_super_super_super_super_super_super_super_long,
  :super_super_super_super_super_super_super_super_super_long => {
    :super_super_super_super_super_super_super_super_long =>
      super_super_super_super_super_super_super_super_long
  }
}

# rubocop:enable Lint/Void
"
`;

exports[`hash.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

{}

{ a: 'a', b: 'b', c: 'c' }

{ a: 'a', b: 'b', c: 'c' }

{ Foo => 1, Bar => 2 }

{
  super_super_super_super_super_super_super_super_long:
    super_super_super_super_super_super_super_super_long,
  super_super_super_super_super_super_super_super_super_long: {
    super_super_super_super_super_super_super_super_long:
      super_super_super_super_super_super_super_super_long
  }
}

# rubocop:enable Lint/Void
"
`;

exports[`hooks.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Style/BeginBlock, Style/EndBlock

BEGIN { p \\"begin\\" }

BEGIN { p \\"begin\\" }

BEGIN {
  super_super_super_super_super_super_super_super_super_super_super_super_long
}

BEGIN {
  super_super_super_super_super_super_super_super_super_super_super_super_long
}

END { p \\"end\\" }

END { p \\"end\\" }

END {
  super_super_super_super_super_super_super_super_super_super_super_super_long
}

END {
  super_super_super_super_super_super_super_super_super_super_super_super_long
}

# rubocop:enable Style/BeginBlock, Style/EndBlock
"
`;

exports[`hooks.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Style/BeginBlock, Style/EndBlock

BEGIN { p 'begin' }

BEGIN { p 'begin' }

BEGIN {
  super_super_super_super_super_super_super_super_super_super_super_super_long
}

BEGIN {
  super_super_super_super_super_super_super_super_super_super_super_super_long
}

END { p 'end' }

END { p 'end' }

END {
  super_super_super_super_super_super_super_super_super_super_super_super_long
}

END {
  super_super_super_super_super_super_super_super_super_super_super_super_long
}

# rubocop:enable Style/BeginBlock, Style/EndBlock
"
`;

exports[`if.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

if a
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

if a
  1
elsif b
  2
end

if a
  super_super_super_super_super_super_super_super_super_super_super_long
else
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

if a
  1
elsif b
  2
elsif c
  3
else
  4
end

unless a
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

# rubocop:disable Style/UnlessElse

unless a
  super_super_super_super_super_super_super_super_super_super_super_long
else
  super_super_super_super_super_super_super_super_super_super_super_super_long
end
# rubocop:enable Style/UnlessElse

if a
  1
end

if super_super_super_super_super_super_super_super_super_super_super_super_long
  1
end

unless a
  1
end

unless super_super_super_super_super_super_super_super_super_super_super_suplong
  1
end

a ? 1 : 2

a ? 2 : 1

a ? 1 : 2

if a
  super_super_super_super_super_super_super_super_super_super_super_long
else
  super_super_super_super_super_super_super_super_super_super_super_super_long
end
"
`;

exports[`if.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

if a
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

if a
  1
elsif b
  2
end

if a
  super_super_super_super_super_super_super_super_super_super_super_long
else
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

if a
  1
elsif b
  2
elsif c
  3
else
  4
end

unless a
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

# rubocop:disable Style/UnlessElse

unless a
  super_super_super_super_super_super_super_super_super_super_super_long
else
  super_super_super_super_super_super_super_super_super_super_super_super_long
end
# rubocop:enable Style/UnlessElse

1 if a

if super_super_super_super_super_super_super_super_super_super_super_super_long
  1
end

1 unless a

unless super_super_super_super_super_super_super_super_super_super_super_suplong
  1
end

a ? 1 : 2

a ? 2 : 1

a ? 1 : 2

if a
  super_super_super_super_super_super_super_super_super_super_super_long
else
  super_super_super_super_super_super_super_super_super_super_super_super_long
end
"
`;

exports[`kwargs.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

def add(alpha:, beta:, gamma: 1, delta: 2)
  alpha + beta + gamma + delta
end

add(:alpha => 1, :beta => 2, :gamma => 3, :delta => 4)

args = { :alpha => 1, :beta => 2, :gamma => 3 }
add(**args, :delta => 4)
"
`;

exports[`kwargs.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

def add(alpha:, beta:, gamma: 1, delta: 2)
  alpha + beta + gamma + delta
end

add(alpha: 1, beta: 2, gamma: 3, delta: 4)

args = { alpha: 1, beta: 2, gamma: 3 }
add(**args, delta: 4)
"
`;

exports[`lambda.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

-> { 1 }

->(a, b, c) { a + b + c }

lambda do
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

lambda do |a, b, c|
  a + b + c + super_super_super_super_super_super_super_super_super_long
end

a.call(1, 2, 3)

a.call(1, 2, 3)

a[]

a[1, 2, 3]

->(a) { a }

-> { 1 }
"
`;

exports[`lambda.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

-> { 1 }

->(a, b, c) { a + b + c }

lambda do
  super_super_super_super_super_super_super_super_super_super_super_super_long
end

lambda do |a, b, c|
  a + b + c + super_super_super_super_super_super_super_super_super_long
end

a.call(1, 2, 3)

a.call(1, 2, 3)

a[]

a[1, 2, 3]

->(a) { a }

-> { 1 }
"
`;

exports[`layout.rb matches expected output for off.json 1`] = `
"1.0

def foobar
  :a

  \\"b\\"
  1
  2

  3
end

2
3
4

5
"
`;

exports[`layout.rb matches expected output for on.json 1`] = `
"1.0

def foobar
  :a

  'b'
  1
  2

  3
end

2
3
4

5
"
`;

exports[`method.rb matches expected output for off.json 1`] = `
"def foo; end

def foo(); end

def foo a; end

def foo(a); end

def self.foo; end

def self.foo(); end

def self.foo a; end

def self.foo(a); end

def foo(super_super_super_super_super_super_super_super_super_super_super_super_super_super_long); end

def foo(a, b, c, super_super_super_super_super_super_super_super_super_super_super_super_super_super_long)
  \\"what\\"
end

def foo(a, b, c = 1, d = 2, *e, f, g, h:, i:, j: 1, k: 2, **l, &block)
  \\"what\\"
end

def foo(a)
  1
end

def foo(*); end

def foo(**); end

foo

foo(1)
foo(1, 2)
foo(1, 2, *abc)
foo(1, 2, *abc, 3, 4)

foo(*bar)
foo(**baz)
foo(&block)

foo(*bar, &block)
foo(**baz, &block)
foo(*bar, **baz, &block)

foo(:h => 1, **bar)
foo(**bar, :h => 1)
foo(:h => 1, **bar, :i => 2)
"
`;

exports[`method.rb matches expected output for on.json 1`] = `
"def foo; end

def foo(); end

def foo a; end

def foo(a); end

def self.foo; end

def self.foo(); end

def self.foo a; end

def self.foo(a); end

def foo(super_super_super_super_super_super_super_super_super_super_super_super_super_super_long); end

def foo(a, b, c, super_super_super_super_super_super_super_super_super_super_super_super_super_super_long)
  'what'
end

def foo(a, b, c = 1, d = 2, *e, f, g, h:, i:, j: 1, k: 2, **l, &block)
  'what'
end

def foo(a)
  1
end

def foo(*); end

def foo(**); end

foo

foo(1)
foo(1, 2)
foo(1, 2, *abc)
foo(1, 2, *abc, 3, 4)

foo(*bar)
foo(**baz)
foo(&block)

foo(*bar, &block)
foo(**baz, &block)
foo(*bar, **baz, &block)

foo(h: 1, **bar)
foo(**bar, h: 1)
foo(h: 1, **bar, i: 2)
"
`;

exports[`next.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

[].each { next }

[].each { next 1 }

[].each { next 1 }
"
`;

exports[`next.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

[].each { next }

[].each { next 1 }

[].each { next 1 }
"
`;

exports[`numbers.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

123
-123
1_123
-543
123_456_789_123_456_789
123.45
1.2e-3
0xaabb # this is actually octal
0o377
-0b1010
0b001_001 # these are CHARs
\\"a\\"
?\\\\C-a
?\\\\M-a
?\\\\M-\\\\C-a

# rubocop:enable Lint/Void
"
`;

exports[`numbers.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

123
-123
1_123
-543
123_456_789_123_456_789
123.45
1.2e-3
0xaabb # this is actually octal
0o377
-0b1010
0b001_001 # these are CHARs
'a'
?\\\\C-a
?\\\\M-a
?\\\\M-\\\\C-a

# rubocop:enable Lint/Void
"
`;

exports[`ranges.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/UselessAssignment

dot2 = 1..2 # dot2
dot3 = 3...4 # dot3

# rubocop:enable Lint/UselessAssignment
"
`;

exports[`ranges.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/UselessAssignment

dot2 = 1..2 # dot2
dot3 = 3...4 # dot3

# rubocop:enable Lint/UselessAssignment
"
`;

exports[`regexp.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

/abc/

%r{a/b/c}

/abc/

/abc/

/abc/

/abc/

/a#{b}c/

/abc/

/abc/

/#{$&}/

# rubocop:enable Lint/Void
"
`;

exports[`regexp.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

/abc/

%r{a/b/c}

/abc/

/abc/

/abc/

/abc/

/a#{b}c/

/abc/

/abc/

/#{$&}/

# rubocop:enable Lint/Void
"
`;

exports[`rescue.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

begin
  1
rescue ArgumentError
  retry
rescue NoMethodError => exception
  puts exception
  redo
rescue SyntaxError, NoMethodError
  2
rescue StandardError
  3
else
  4
ensure
  5
end

begin
  a
rescue StandardError
  nil
end
"
`;

exports[`rescue.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

begin
  1
rescue ArgumentError
  retry
rescue NoMethodError => exception
  puts exception
  redo
rescue SyntaxError, NoMethodError
  2
rescue StandardError
  3
else
  4
ensure
  5
end

begin
  a
rescue StandardError
  nil
end
"
`;

exports[`return.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

def foo
  if a
    return
  end
end

def bar
  if a
    return 1
  end
end

def baz
  if a
    return 1
  end
end

def qux
  if a
    return 1, 2
  end
end
"
`;

exports[`return.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

def foo
  return if a
end

def bar
  return 1 if a
end

def baz
  return 1 if a
end

def qux
  return 1, 2 if a
end
"
`;

exports[`strings.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

\\"\\"

\\"abc\\"

\\"abc\\"

\\"#{abc} abc\\"

\\"abc #{de} fghi #{jkl} mno\\"

\\"abc\\" \\\\
  \\"def\\" \\\\
  \\"ghi\\"

\\"abc #{\\"abc #{abc} abc\\"} abc\\"

{ \\"a\\" => 1 }

{ \\"a #{a}\\" => 1 }

:\\"abc#{abc}abc\\"

\`abc\`

\`
  super_super_super_super_super_super_super_super_super_super_super_super_su_long
\`

# rubocop:enable Lint/Void
"
`;

exports[`strings.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

# rubocop:disable Lint/Void

''

'abc'

'abc'

\\"#{abc} abc\\"

\\"abc #{de} fghi #{jkl} mno\\"

'abc' \\\\
  'def' \\\\
  'ghi'

\\"abc #{\\"abc #{abc} abc\\"} abc\\"

{ 'a' => 1 }

{ \\"a #{a}\\" => 1 }

:\\"abc#{abc}abc\\"

\`abc\`

\`
  super_super_super_super_super_super_super_super_super_super_super_super_su_long
\`

# rubocop:enable Lint/Void
"
`;

exports[`super.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

super

super 1

super(1)

super 1, 2

super(1, 2)
"
`;

exports[`super.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

super

super 1

super(1)

super 1, 2

super(1, 2)
"
`;

exports[`while.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

while a
  1
end

while a
  1
end

while super_super_super_super_super_super_super_super_super_super_super_long
  1
end

while super_super_super_super_super_super_super_super_super_super_super_supelong
  1
end

until a
  1
end

until a
  1
end

until super_super_super_super_super_super_super_super_super_super_super_long
  1
end

until super_super_super_super_super_super_super_super_super_super_super_supelong
  1
end
"
`;

exports[`while.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

1 while a

1 while a

1 while super_super_super_super_super_super_super_super_super_super_super_long

while super_super_super_super_super_super_super_super_super_super_super_supelong
  1
end

1 until a

1 until a

1 until super_super_super_super_super_super_super_super_super_super_super_long

until super_super_super_super_super_super_super_super_super_super_super_supelong
  1
end
"
`;

exports[`yield.rb matches expected output for off.json 1`] = `
"# frozen_string_literal: true

[1, 2, 3].each do |i|
  yield

  yield i

  yield(i)

  yield i, 2

  yield(i, 2)

  yield
end
"
`;

exports[`yield.rb matches expected output for on.json 1`] = `
"# frozen_string_literal: true

[1, 2, 3].each do |i|
  yield

  yield i

  yield(i)

  yield i, 2

  yield(i, 2)

  yield
end
"
`;
